crypto/merkle.go

```diff
@@ -71,26 +71,26 @@ func ReaderMerkleRoot(r io.Reader) (h Hash, err error) {
 }
 
 // BuildReaderProof will build a storage proof when given a reader.
-func BuildReaderProof(r io.Reader, proofIndex uint64) (base [SegmentSize]byte, hashSet []Hash, err error) {
+func BuildReaderProof(r io.Reader, proofIndex uint64) (base []byte, hashSet []Hash, err error) {
 	_, proofSet, _, err := merkletree.BuildReaderProof(r, NewHash(), SegmentSize, proofIndex)
 	if err != nil {
 		return
 	}
 	// convert proofSet to base and hashSet
-	copy(base[:], proofSet[0])
+	base = proofSet[0]
 	hashSet = make([]Hash, len(proofSet)-1)
-	for i, proof := range proofSet[1:] {
-		copy(hashSet[i][:], proof)
+	for i := range proofSet[1:] {
+		copy(hashSet[i][:], proofSet[i+1])
```

Comment:
ok, fixed.

I don't like making extra memory copies when you don't have to.

But these might actually be efficient because they just use the same memory repeatedly. I guess it's something we could benchmark.